cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/SparkUdt.java

```diff
@@ -0,0 +1,179 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.cassandra.spark.data.converter.types.complex;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.apache.cassandra.bridge.BigNumberConfig;
+import org.apache.cassandra.spark.data.CqlField;
+import org.apache.cassandra.spark.data.converter.SparkSqlTypeConverter;
+import org.apache.cassandra.spark.data.converter.types.SparkType;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.StructField;
+
+import java.nio.ByteBuffer;
+
+public class SparkUdt implements SparkType
+{
+    private final SparkSqlTypeConverter converter;
+    private final CqlField.CqlUdt udt;
+    private final List<SparkType> sparkTypes;
+
+    public SparkUdt(SparkSqlTypeConverter converter, CqlField.CqlUdt udt)
+    {
+        this.converter = converter;
+        this.udt = udt;
+        this.sparkTypes = udt.fields().stream()
+                             .map(CqlField::type)
+                             .map(converter::toSparkType)
+                             .collect(Collectors.toList());
+    }
+
+    public CqlField field(int position)
+    {
+        return this.udt.field(position);
+    }
+
+    @Override
+    public DataType dataType(BigNumberConfig bigNumberConfig)
+    {
+        return DataTypes.createStructType(
+        udt.fields().stream()
+           .map(field -> DataTypes.createStructField(field.name(),
+                                                     converter.sparkSqlType(field.type(), bigNumberConfig),
+                                                     true))
+           .toArray(StructField[]::new)
+        );
+    }
+
+    @Override
+    public Object sparkSqlRowValue(GenericInternalRow row, int position)
+    {
+        final InternalRow struct = row.getStruct(position, size());
+        return IntStream.range(0, size())
+                        .boxed()
+                        .collect(Collectors.toMap(
+                        index -> field(index).name(),
+                        index -> sparkType(index).toTestRowType(struct.get(index, sparkType(index).dataType())
+                        )));
+    }
+
+    @Override
+    public Object sparkSqlRowValue(Row row, int position)
+    {
+        final Row struct = row.getStruct(position);
```

Comment:
NIT unnecessary final modifier
```suggestion
        Row struct = row.getStruct(position);
```